core/interop/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+//! Interop utilities between Boa and its host.
+use std::cell::RefCell;
+
+use boa_engine::module::SyntheticModuleInitializer;
+use boa_engine::{Context, JsString, JsValue, Module, NativeFunction};
+
+pub mod loaders;
+
+/// A trait to convert a type into a JS module.
+pub trait IntoJsModule {
+    /// Converts the type into a JS module.
+    fn into_js_module(self, context: &mut Context) -> Module;
+}
+
+impl<T: IntoIterator<Item = (JsString, NativeFunction)> + Clone> IntoJsModule for T {
+    fn into_js_module(self, context: &mut Context) -> Module {
+        let (names, fns): (Vec<_>, Vec<_>) = self.into_iter().unzip();
+        let exports = names.clone();
+
+        Module::synthetic(
+            exports.as_slice(),
+            unsafe {
+                SyntheticModuleInitializer::from_closure(move |module, context| {
+                    for (name, f) in names.iter().zip(fns.iter()) {
+                        module
+                            .set_export(name, f.clone().to_js_function(context.realm()).into())?;
+                    }
+                    Ok(())
+                })
+            },
+            None,
+            context,
+        )
+    }
+}
+
+/// A trait to convert a type into a JS function.
+pub trait IntoJsFunction {
+    /// Converts the type into a JS function.
+    fn into_js_function(self, context: &mut Context) -> NativeFunction;
+}
+
+impl<T: FnMut() + 'static> IntoJsFunction for T {
```

Comment:
This is probably UB, or at least it could cause a memory leak if `T` captures `Gc` types within it.

The reason we require `Copy` on `NativeFunction::from_copy_closure` is because Rust currently doesn't allow inspecting the captured variables of a closure. This makes it impossible to trace garbage collected variables that are captures of a closure. However, `Gc` is only `Clone`, making it impossible to be inside a `Fn() + Copy`.

Reply:
Okay I see it now. That makes using closures very hard :( I need to experiment a bit.

Reply:
I think this could probably be fine if the unsafe marker is lifted to the user, so that the user has to ensure the closures don't contain `Gc` types. This would probably imply splitting the current trait into two: a safe variant implementable by `Fn + Copy` and an unsafe variant implementable by `FnMut`.

Reply:
Okay, I've added unsafe to both the trait and the method, will implement a safe version separately. PTAL.

Reply:
Yep, seems fine to start with an unsafe trait for now. We can always adjust the design after designing the macros first.